### ISSUE TITLE
syscall: Older versions of the Linux kernel return the `EBADF` error instead of `ENOSYS` when calling `faccessat2`

### ISSUE THREAD

**Issue by Mac-Genius** [golang/go#65243]
### Go version

go version go1.21.6 linux/amd64

### Output of `go env` in your module/workspace:

```shell
GO111MODULE=''
GOARCH='AMD64'
GOCACHE='/home/user/.cache/go-build'
GOENV='/home/user/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS=''
GOINSECURE=''
GOMODCACHE='/home/user/.config/go/env'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/user/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/home/user/tools/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/home/user/tools/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.21.6'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='0'
GOMOD='/dev/null'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build1105951330=/tmp/go-build -gno-record-gcc-switches'
```

### What did you do?

I compiled and ran the following program on my machine: https://go.dev/play/p/FBV_Gfekn4y

### What did you see happen?

The result of running the program:

```bash
[user@computer] ./test
exec: "git" executable file not found in $PATH
```

### What did you expect to see?

It should have printed the location of the `git` executable. Here is the output of the `PATH` environment variable:

```bash
[user@computer] echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

This seems to be a specific issue with 1.20 and newer versions of Go. When I build and run the program using `go1.19.13`:

```bash
[user@computer] ./test
/usr/bin/git
```

It correctly outputs the path of the `git` executable.

The issue seems to stem from the function `findExecutable` in `src/os/exec/lp_unix.go` using the `unix.Eaccess` function. This function calls `Faccessat` found in `src/syscall/syscall_linux.go` which in turn calls `faccessat2`. On older versions of the Linux kernel it appears that the OS returns the `EBADF` error for unimplemented syscalls instead of `ENOSYS` or `EPERM`.

This is running on a old custom flavor of CentOS 7. Here is the output of uname:

```bash
[user@computer] uname -a
Linux computer 4.1.37-Custom #1 SMP PREEMPT Wed Mar 14 15:29:41 EDT 2018 x86_64 x86_64 x86_64 GNU/Linux
```

Here is a test program I wrote to test what error was returned from `faccessat`: https://go.dev/play/p/bhLaQ53EU6P

Output:
```bash
[user@computer] ./test
1: <nil>
2: bad file descriptor
3: bad file descriptor
```

On newer kernels it appears to return the expected error:

```bash
[user@computer] uname -a
Linux computer 4.15.0-213-generic #224-Ubuntu SMP Mon Jun 19 13:30:12 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
[user@computer] ./test
1: <nil>
2: function not implemented
3: function not implemented
```

TLDR: `faccessat2` returns the `EBADF` error on older Linux kernels instead of `ENOSYS` or `EPERM`, resulting in `syscall.Faccessat` to always fail on these older machines. This in turn causes the `exec.LookPath` function to always fail since the `EBADF` error is always returned.

**Comment by bcmills** [comment]
@Mac-Genius, it doesn't make sense for the 4.1.37 kernel to return `EBADF` for the `faccessat2` syscall, since that syscall didn't even exist until Linux 5.8. That version of the kernel has no way of knowing that the syscall number even corresponds to something that operates on a file descriptor!

In #57393 we observed failures on Android due to a bad `seccomp` filter on that platform. Is it possible that your CentOS system has a similar `seccomp` issue? If so, can you fix the behavior of the program by adjusting your `seccomp` policy to return a more sensible error code?

**Comment by bcmills** [comment]
> This is running on a old custom flavor of CentOS 7.

Is it possible that your old custom kernel defined syscall number 439 to implement some custom system call that didn't make it to the mainline kernel? (Does the `syscall.tbl` in your kernel source define anything for syscall number 439?)

**Comment by Mac-Genius** [comment]
Hi @bcmills, you are indeed correct, it looks like this custom image implements a syscall for 439. Is there a way to work around this issue without having to disable this syscall? I do not have the full source for this image, it was provided by a vendor.

**Comment by bcmills** [comment]
Unfortunately, I think your options are probably to block the syscall (perhaps using a `seccomp` filter?) or patch the Go standard library to match the semantics of your patched kernel.

**Comment by Mac-Genius** [comment]
I will most likely move forward with a patch for Go deployed on this particular OS. Thank you for your help!
